mailbox/plugin/spamassassin/src/main/java/org/apache/james/mailbox/spamassassin/SpamAssassin.java

```diff
@@ -18,9 +18,11 @@
  ****************************************************************/
```

Comment:
How does the work performed in https://github.com/linagora/james-project/pull/3281/commits/65c2512d7bc2087a36d3db59cfa08ce52af45b5d `JAMES-3167 Reactify SpamAssassin ` relates to https://github.com/linagora/james-project/pull/3253 and especially to https://github.com/linagora/james-project/pull/3253/commits/b57b22b52925b7d2fc87d0d4c1c3afa197d50964 `JAMES-3128 use reactor netty TCPClient to call spamassassin` ?

Reply:
#3253 allow an unbounded concurrency, while my commit give a way to manage it.

You are right, but it would require a global policy which states to not include quick improvements. But it would be detrimental IMO.

Reply:
> #3253 allow an unbounded concurrency, while my commit give a way to manage it.

Do you think the concurrency level we allow can be configurable?

I still think that enforcing concurrency should be done at the eventBus level (as it is the one planning the computation) rather than deep down inside a specific computation.

> You are right, but it would require a global policy which states to not include quick improvements. But it would be detrimental IMO.

Sorry I don't understand this sentence.

I had been told that splitting pull requests was a good strategy to actually get quick improvments faster (like your first commit) while we review and discuss the strategy to manage SpamAssassin concurrency.

Reply:
> Do you think the concurrency level we allow can be configurable?

I do not get it.

> I still think that enforcing concurrency should be done at the eventBus level (as it is the one planning the computation) rather than deep down inside a specific computation.

I am not sure it would be relevant here, you want to reduce the pressure on a backend, given the knowledge of this backend's usage to the eventbus would create a coupling and lead to errors.

> I had been told that splitting pull requests was a good strategy to actually get quick improvments faster (like your first commit) while we review and discuss the strategy to manage SpamAssassin concurrency.

Sure, but we actually do not do it for small changes.

Reply:
>  Sure, but we actually do not do it for small changes.

That is not true.

And ` 61 files changed, 1432 insertions(+), 1468 deletions(-)` is no  small changes.

Also the to last commit don't relate to the title of the PR.

> Do you think the concurrency level we allow can be configurable?

What if as an admin I payed a super big fast machine for SpamAssassin and wants to process 2, 5, 10 messages in parallel?

Maybe exposing this as a capability in a configuration file could be nice...

>>  I still think that enforcing concurrency should be done at the eventBus level (as it is the one planning the computation) rather than deep down inside a specific computation.

> I am not sure it would be relevant here, you want to reduce the pressure on a backend, given the knowledge of this backend's usage to the eventbus would create a coupling and lead to errors.

Well that `Backend` providing spam feedback is accessed via a single `mailboxListener`. I'm expecting an admin to know that Spam feedback happens via a mailboxListener.

In my opinion you proposal adds yet-another place where to specify some execution parameters, that duplicates the knowledge of other components.

Finally, we miss a powerfull tool: reuse. For example, what if I want to rate limit `elasticSearch indexation` too? (Would it makes sense?) Should I do it with Reactor wizardry at the backend level?

Reply:
>> Sure, but we actually do not do it for small changes.
>
> That is not true.
>
> And 61 files changed, 1432 insertions(+), 1468 deletions(-) is no small changes.

Sure, what is the percentage dedicated to SA?

> Also the to last commit don't relate to the title of the PR.

You have my opinion here: https://github.com/linagora/james-project/pull/3281#discussion_r430213956

> > Do you think the concurrency level we allow can be configurable?
>
> What if as an admin I payed a super big fast machine for SpamAssassin and wants to process 2, 5, 10 messages in parallel?
>
> Maybe exposing this as a capability in a configuration file could be nice...

https://github.com/linagora/james-project/pull/3281/commits/c02b8417dbc131a5b71867c14056782d01d586f9#diff-6bf4985f37d23a58fb53b49ada7294bcR39

> Well that Backend providing spam feedback is accessed via a single mailboxListener. I'm expecting an admin to know that Spam feedback happens via a mailboxListener.

Sure, the admin, not the unrelated component: https://blog.ndepend.com/solid-design-the-dependency-inversion-principle-dip/

> In my opinion you proposal adds yet-another place where to specify some execution parameters, that duplicates the knowledge of other components.

You do not duplicate it,  you segregate it in order to avoid unnecessary coupling.

> Finally, we miss a powerfull tool: reuse. For example, what if I want to rate limit elasticSearch indexation too? (Would it makes sense?) Should I do it with Reactor wizardry at the backend level?

Reusage (DRY) is often a bad practice, especially when it creates coupling.

Reply:
> You have my opinion here: #3281 (comment)
> "#3253 allow an unbounded concurrency, while my commit give a way to manage it."

How can I understand such a message?

I need literraly 10 minutes, and a review of @remk work to understand that even with a bounded event processing rate, as an Added event can relate to several event my assumption was false.

Great. I would suggest:

 - `flatMap(operation, 1)` on spam reporting in @remk work to keep concurrency control
 - And with the invarant `event concurrency ~ spamassassin concurrency` restored, nothing prevent us to rate limit spamassassin events.

>> Finally, we miss a powerfull tool: reuse. For example, what if I want to rate limit elasticSearch indexation too? (Would it makes sense?) Should I do it with Reactor wizardry at the backend level?

> Reusage (DRY) is often a bad practice, especially when it creates coupling.

Concurrency control is a **feature** I expect from the EventBus.

I would say 'it should be thanks to back-pressure self adaptative', but removing the hard cpoded constant QOS  Execution rate in `GroupRegistration` is IMO a nice to have thing to have, and addresses your concerns too.

Reply:
> How can I understand such a message?

You asked how it was different, I legitimately assumed that you know it.

> Great. I would suggest:

> * flatMap(operation, 1) on spam reporting in @remk work to keep concurrency control
> * And with the invarant event concurrency ~ spamassassin concurrency restored, nothing prevent us to rate limit spamassassin events.

That would be worse in these case:

 * Rate limiting is < concurrency level
 * Rate limiting is > concurrency level

if you're SA can handle 4 concurrent requests, and you have only two event at a time, you will be slower than you could for example.

> Concurrency control is a feature I expect from the EventBus.

Then you will couple high-level abstraction to low-level behaviors.

> I would say 'it should be thanks to back-pressure self adaptative', but removing the hard cpoded constant QOS Execution rate in GroupRegistration is IMO a nice to have thing to have, and addresses your concerns too.

My goal is to optimize backend concurrency, not to limit the processessing rate.

Reply:
>> How can I understand such a message?

>  You asked how it was different, I legitimately assumed that you know it.

I might then not be clever as you think I am...

> For example: if you're SA can handle 4 concurrent requests, and you have only two events at a time, you will be slower than necessary.

Who cares. We are in an asynchronous process, with limited importance. We don't care if a given event takes 100 ms or 500ms to be processed, we care about event throughtput, and both proposal are similar from that perspective.

>> Concurrency control is a feature I expect from the EventBus.

> Then you will couple high-level abstraction to low-level behaviors.

>> I would say 'it should be thanks to back-pressure self adaptative', but removing the hard cpoded constant QOS Execution rate in GroupRegistration is IMO a nice to have thing to have, and addresses your concerns too.

> My goal is to optimize backend concurrency, not to limit the processessing rate.

Well, I don't see us making any progress here. What do other people think about this topic?

Anyway, feel free to move forward, I don't need to agree with all the lines contributed to the software I work on.

If we are unwilling to come back to the design table, then feel free to ignore my comments. I will then let another Robusta do the cross review for that PR. @Arsnael maybe?

Reply:
>> You asked how it was different, I legitimately assumed that you know it.
>
> I might then not be clever as you think I am...

It's not the point.

> Who cares. We are in an asynchronous process, with limited importance. We don't care if a given event takes 100 ms or 500ms to be processed, we care about event throughtput, and both proposal are similar from that perspective.

We care, let's take the cluster point-of-view, we have four nodes and four events, if each node gets one event, each events will be processed in 500 ms instead of 100 ms.
Our goal are very differents, my PR make a step in the direction of having a stable event state, while limiting event rate may result in sub-optimal backend usage.

Reply:
> Given the context of that Pull request, I would suggest to only keep the first commit...

I did not follow the technical arguments, but I wanted to explicit this point.

The main goal we have all to achieve is readability. If a PR is written by one person, it will be read by at least 3 people, and surely the code will be read several times more in the future.

So there is no hard rule on putting more or less related changes in a PR. But if a PR is hard to review because there are too many unrelated changes in it, it have to be split. So if a reviewer has a comment telling that it's hard, it must be split.

In this case @chibenwa I did not get exactly your concern: do you wish to split this PR because it's too hard, or just by principle?

Reply:
I agree 100% with reactifying ACL updates (1st commit & PR title). (and I agree approuving it any time, even if the changeset is big)

However I have issues (raised above) with the two other commits, relative to SpamAssassin, so not related either to the first commit nor the PR title.

So I wonder why is this integrated in this work?

Reply:
ok, for me that seems to be a good reason to split this PR